pyro/contrib/gp/kernels/coregionalize.py

```diff
@@ -9,50 +10,86 @@
+    :math:`k(x,z) = x^T (C C^T + D) z` where :math:`C` is an
+    ``input_dim``-by-``rank`` matrix and typically ``rank < input_dim``,
+    and ``D`` is a diagonal matrix.
 
     This generalizes the
     :class:`~pyro.contrib.gp.kernels.dot_product.Linear` kernel to multiple
-    features. The typical use case is for modeling correlations among outputs
-    of a multi-output GP, where outputs are coded as distinct data points with
-    one-hot coded features denoting which output each datapoint represents.
+    features with a low-rank-plus-diagonal weight matrix. The typical use case
+    is for modeling correlations among outputs of a multi-output GP, where
+    outputs are coded as distinct data points with one-hot coded features
+    denoting which output each datapoint represents.
+
+    If only ``rank`` is specified, the kernel ``(C C^T + D)`` will be
+    randomly initialized to a matrix with expected value the identity matrix.
 
     References:
 
     [1] Mauricio A. Alvarez, Lorenzo Rosasco, Neil D. Lawrence (2012)
         Kernels for Vector-Valued Functions: a Review
 
     :param int input_dim: Number of feature dimensions of inputs.
+    :param int rank: Optional rank. This is only used if ``components`` is
+        unspecified. If neigher ``rank`` nor ``components`` is specified,
+        then ``rank`` defaults to ``input_dim``.
     :param torch.Tensor components: An optional ``(input_dim, rank)`` shaped
         matrix that maps features to ``rank``-many components. If unspecified,
         this will be randomly initialized.
-    :param int rank: Optional rank. This is only used if ``components`` is
-        unspecified. If unspecified, ``rank`` defaults to ``input_dim``.
+    :param torch.Tensor diagonal: An optional vector of length ``input_dim``.
+        If unspecified, this will be set to constant ``1 - rank / input_dim`` if
+        ``rank < input_dim`` otherwise zero.
     :param list active_dims: List of feature dimensions of the input which the
         kernel acts on.
     :param str name: Name of the kernel.
     """
 
-    def __init__(self, input_dim, components=None, rank=None, active_dims=None, name="coregionalize"):
+    def __init__(self, input_dim, rank=None, components=None, diagonal=None, active_dims=None, name="coregionalize"):
         super(Coregionalize, self).__init__(input_dim, active_dims, name)
+
+        # Add a low-rank kernel with expected value torch.eye(input_dim, input_dim) * rank / input_dim.
         if components is None:
-            if rank is None:
-                rank = input_dim
-            components = torch.randn(input_dim, rank) / rank ** 0.5
-        if components.dim() != 2 or components.shape[0] != input_dim:
-            raise ValueError("Expected region.shape == ({},rank), actual {}".format(input_dim, components.shape))
+            rank = input_dim if rank is None else rank
+            components = torch.randn(input_dim, rank) / input_dim ** 0.5
```

Comment:
I think that the expectation of coregionalization matrix B is Identity. The `components` will transforms (one-hot version of) the set {1, ..., D} to the row vectors {w1, ..., wD). Hence, diag(B) = [w1.w1, w2.w2, ..., wD.wD]. So we expect E(wi.wi) = 1. This is achieved if we initialize `components = randn(...) / rank ** 0.5`.

Reply:
I'd like the expectation of the entire kernel `C C^T + D` to be the identity. That's why I've split the identity in to the `rank / input_dim` + `(1 - rank / input_dim)` parts corresponding to `E[ C C^T ]` and `D`, respectively.

Reply:
Got it! If you want to keep the diagonal part, then how about making them `rank / (input_dim + rank)`, `(1 - rank / input_dim + rank)`? When using diagonal, then this diagonal term will give the independent info for outputs, hence it should have a bigger weight. This way, we will have `diagonal` for case `input_dim <= rank`, so the implementation will be simplified.

Reply:
> so the implementation will be simplified

Hmm I'm not seeing how the implementation would simplify the code. IIUC your suggestion would involve replacing the `torch.randn(input_dim, rank) / rank` with `torch.randn(input_dim, rank) / (input_dim + rank)` and replacing `1 - rank / input_dim` with `1 - rank / (input_dim + rank)`. Could you explain?

Reply:
I think that in that case, we can remove the situation `diagonal = None` in the implementation because `diagon > 0` regardless rank >= input_dim.